src/projection.jl

```diff
@@ -181,9 +192,19 @@ end
 # If we don't have a more specialized `ProjectTo` rule, we just assume that there is
 # no structure worth re-imposing. Then any array is acceptable as a gradient.
 
-# For arrays of numbers, just store one projector:
-function ProjectTo(x::AbstractArray{T}) where {T<:Number}
-    return ProjectTo{AbstractArray}(; element=_eltype_projectto(T), axes=axes(x))
+# For arrays of numbers, just store one projector, and construct it without branches:
+ProjectTo(x::AbstractArray{<:Number}) = _array_projectto(x, axes(x))
+function _array_projectto(x::AbstractArray{T,N}, axes::NTuple{N,<:Base.OneTo{Int}}) where {T,N}
+    element = _eltype_projectto(T)
+    S = project_type(element)
+    # Fastest path: N means they are OneTo, hence reshape can be skipped
+    return ProjectTo{AbstractArray{S,N}}(; element=element, axes=axes)
```

Comment:
This is slightly weird, but my idea for avoiding the `if axes(dx) == projector.axes` branch is to include the `N` only when the axes are `Base.OneTo`. If `ndims(dx)` matches, and its eltype matches, then it can pass through just by dispatch.

This means it won't check for quite as many size mismatches. But it will still `reshape` for OffsetArrays, SArrays, etc, those will go the "slow path" as before.